src/Elmish/Hooks/UseEffect.purs

```diff
@@ -26,24 +31,80 @@ import Elmish.Hooks.Type (Hook, mkHook, uniqueNameFromCurrentCallStack, uniqueNa
 -- |   pure $ H.fragment $ todoView <$> todos
 -- | ```
 useEffect :: Aff Unit -> Hook Unit
-useEffect = useEffect' identity uniqueNameFromCurrentCallStack
+useEffect aff =
+  useEffect_ name identity unit $ const aff
+  where
+    name = uniqueNameFromCurrentCallStack { skipFrames: 3 }
+
+-- | This is like `useEffect`, but allows passing a value which, when it
+-- | changes, will trigger the effect to run again. E.g.:
+-- |
+-- | ```purs
+-- | view :: ReactElement
+-- | view = withHooks do
+-- |   count /\ setCount <- useState 0
+-- |
+-- |   useEffect' count \c -> liftEffect $
+-- |     HTMLDocument.setTitle ("You clicked " <> show c <> " times") =<< document =<< window
+-- |
+-- |   pure H.button_ "" { onClick: setCount $ count + 1 } "Click me"
+-- | ```
+useEffect' :: forall a. Eq a => a -> (a -> Aff Unit) -> Hook Unit
+useEffect' deps = \aff ->
+  useEffect_ name identity deps aff
+  where
+    name = uniqueNameFromCurrentCallStack { skipFrames: 3 }
 
--- | A version of `useEffect` that logs messages, state changes, render times,
--- | and info from the name-generating function. Intended to be used with
--- | qualified imports: `UseEffect.traced`.
+-- | A version of `useEffect` that logs info from the name-generating function.
+-- | Intended to be used with qualified imports: `UseEffect.traced`.
 traced :: DebugWarning => Aff Unit -> Hook Unit
-traced = useEffect' withTrace uniqueNameFromCurrentCallStackTraced
+traced aff =
+  useEffect_ name withTrace unit $ const aff
+  where
+    name = uniqueNameFromCurrentCallStackTraced { skipFrames: 3 }
+
+-- | A version of `useEffect'` that logs info from the name-generating function.
+-- | Intended to be used with qualified imports: `UseEffect.traced'`.
+traced' :: forall a. DebugWarning => Eq a => a -> (a -> Aff Unit) -> Hook Unit
+traced' deps = \aff ->
+  useEffect_ name withTrace deps aff
+  where
+    name = uniqueNameFromCurrentCallStackTraced { skipFrames: 3 }
 
-useEffect' ::
-  (ComponentDef Void Unit -> ComponentDef Void Unit)
-  -> ({ skipFrames :: Int } -> ComponentName)
-  -> Aff Unit
+useEffect_ :: forall a.
+  Eq a
+  => ComponentName
+  -> (ComponentDef a a -> ComponentDef a a)
+  -> a
+  -> (a -> Aff Unit)
   -> Hook Unit
-useEffect' f genName init =
+useEffect_ name = \f deps init ->
   mkHook name \render -> f
-    { init: forkVoid init
-    , update: \_ msg -> absurd msg
-    , view: \_ _ -> render unit
+    { init: do
+        forkVoid $ init deps
+        pure deps
+    , update: \_ newDeps -> do
+        forkVoid $ init newDeps
+        pure newDeps
+    , view: \_ dispatch ->
+        lifeCycles
+          { componentDidUpdate: dispatch <?| \prevDeps ->
+              if deref prevDeps /= deps then
+                Just deps
+              else
+                Nothing
+          , deps: ref deps
```

Comment:
I don't think you need to use `ref` here. It's just a protection from accidentally mixing up values in JS, and here that would be very unlikely. 

I actually want to remove the `ref` completely, I don't think it's useful in practice.

Reply:
But wouldn't I then need a `CanPassToJavaScript` constraint?

Reply:
Yeah, if I remove all the `ref`s, I get: `No type class instance was found for Elmish.Foreign.CanReceiveFromJavaScript a5`. So I would have to add a constraint to what can be passed as `deps` in `useEffect_`.